[IMP] cetmix_tower_server: Access to server fields

GabbasovDinar opened the pull request:
Limit access to the following fields of the `cx.tower.server` ​model to Manager and Root:
- OS
- IP v4 Address
- IP v6 Address
- SSH username 
- SSH Port
- SSS auth mode
- SSH key
- Use sudo

Task: 4116

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **Security Enhancements**
	- Added group-based access control for server connection settings.
	- Restricted sensitive server configuration fields to the manager group.
	- Made SSH password field optional.

- **User Interface**
	- Introduced a new kanban view for the manager group with restricted field access.
	- Updated search view to reflect new access controls and visibility based on user roles.

- **Documentation**
	- Added note about user access levels in the configuration guide.

- **Testing**
	- Enhanced test coverage for user access and command execution, including sensitive variables.
	- Removed a test for unique constraint validation on variable options.
	- Streamlined test cases by removing unnecessary mocks and directly testing functionality.
	- Added new tests for verifying user access to files with sensitive variables.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

GabbasovDinar committed:
[IMP] cetmix_tower_server: Access to server fields

Limit access to the following fields of the cx.tower.server ​omdel to Manager and Root:

- OS
- IP v4 Address
- IP v6 Address
- SSH username
- SSH Port
- SSS auth mode
- SSH key
- Use sudo

- Fix user access error when uploading a file
- Update tests

Task: 4116

GabbasovDinar commented on cetmix_tower_server/models/cx_tower_server.py:
```diff
@@ -279,15 +279,29 @@ class CxTowerServer(models.Model):
+    ip_v6_address = fields.Char(
+        string="IPv6 Address", groups="cetmix_tower_server.group_manager"
+    )
+    ssh_port = fields.Char(
+        string="SSH port",
+        required=True,
+        default="22",
+        groups="cetmix_tower_server.group_manager",
+    )
```
user cannot create records for this model, only read, so it's ok

GabbasovDinar commented on cetmix_tower_server/models/cx_tower_server.py:
```diff
@@ -279,15 +279,29 @@
+    ssh_username = fields.Char(
+        string="SSH Username", required=True, groups="cetmix_tower_server.group_manager"
+    )
+    ssh_password = fields.Char(
+        string="SSH Password", groups="cetmix_tower_server.group_manager"
+    )
```
user cannot create records for this model, only read, so it's ok

ivs-cetmix commented on cetmix_tower_server/views/cx_tower_server_view.xml:
```diff
@@ -360,7 +372,7 @@
                 <field name="reference" />
                 <field name="status" />
-                <field name="os_id" />
+                <field name="os_id" groups="cetmix_tower_server.group_manager" />
```
Question: why do we need to add `group` attribute to the view definition?

GabbasovDinar replied:
I agree, I will fix it

ivs-cetmix reviewed (changes requested):
We have changed the flow so need to update the test coverage accordingly.
Eg we need to ensure that user with group "Tower/User" can still run SSH commands with no issues.

ivs-cetmix commented on cetmix_tower_server/tests/test_command.py:
```diff
@@ -529,6 +529,16 @@ def test_user_access_rule(self):
         command_name = test_command_1_as_bob.name
         self.assertEqual(command_name, "Test command", msg="Must return 'Test command'")
+        # Check that user with "cetmix_tower_server.group_user" can execute ssh command
+        test_command.write(
```
I think we need to update [this method](https://github.com/cetmix/cetmix-tower/blob/c428b9e07d021f38d09889d7cf898b4c9d005d3c/cetmix_tower_server/tests/common.py#L219) too. And add some read operations to ensure that they don't trigger any access issues.
However let's merge https://github.com/cetmix/cetmix-tower/pull/181 first

ivs-cetmix reviewed (changes requested):
I think we need

ivs-cetmix commented on cetmix_tower_server/wizards/cx_tower_command_execute_wizard.py:
```diff
@@ -53,6 +53,7 @@ class CxTowerCommandExecuteWizard(models.TransientModel):
     any_server = fields.Boolean(default=True)
     rendered_code = fields.Text(
         compute="_compute_rendered_code",
+        compute_sudo=True,
```
Shouldn't we add `compute_sudo` [here](https://github.com/cetmix/cetmix-tower/blob/3c3f48a4767ca4d83b97913fa3de7914db01aa28/cetmix_tower_server/models/cx_tower_file.py#L119) too?

GabbasovDinar replied:
yes, thanks!

Aldeigja reviewed (approved):
functional LGTM

GabbasovDinar committed:
[REF] cetmix_tower_server: Tests

Use MagicMock for mocking SSH and file operations.
This is done to make the test process more clean and clear.

Task: 4116

GabbasovDinar commented on cetmix_tower_server/models/cx_tower_file.py:
```diff
@@ -644,7 +647,7 @@ def _process(self, action, raise_error=False):
                 else:
                     return False
-                file.server_response = "ok"
+                file.sudo().server_response = "ok"
```
can be a some command that allows user to start loading a file, by default the user does not have access to modify the file, but the command must be executed, so this method allows you to update techical data on the file when the command is executed.

ivs-cetmix commented on cetmix_tower_server/tests/test_file.py:
```diff
@@ -392,3 +392,33 @@ def test_file_with_secret_key(self):
+        self.server_test_1.message_subscribe([self.user_bob.partner_id.id])
+        # Upload file to server
+        file.with_user(self.user_bob).action_push_to_server()
+        self.assertEqual(file.server_response, "ok")
```
What if `server_response` was already "ok" before we pushed the file? And our push operation simply didn't check it.
Probably we need to set `server_response` to some other value before running the test.

GabbasovDinar replied:
done

ivs-cetmix reviewed (changes requested):
Note: no need to add `refracting` in the commit message again because there is a [REF] tag already. So something like:

```
[REF] cetmix_tower_server: Tests

Use MagicMock for mocking SSH and file operations.
This is done to make the test process more clean and clear.
```

ivs-cetmix reviewed (approved):
Code review LGTM

CetmixGitDrone commented:
This PR has the `approved` label and has been created more than 5 days ago. It should therefore be ready to merge by a maintainer (or a PSC member if the concerned addon has no declared maintainer). 🤖

hoangtrann reviewed (approved):
Code LGTM

ivs-cetmix commented:
/ocabot merge nobump

CetmixGitDrone commented:
This PR looks fantastic, let's merge it!
Prepared branch [14.0-dev-ocabot-merge-pr-182-by-ivs-cetmix-bump-nobump](https://github.com/cetmix/cetmix-tower/commits/14.0-dev-ocabot-merge-pr-182-by-ivs-cetmix-bump-nobump), awaiting test results.

CetmixGitDrone commented:
Congratulations, your PR was merged at 9e8bc1fdb5ec4a3eb5bc8bfef0204500be029b3f. Thanks a lot for contributing to cetmix. ❤️
